regression/cuda/test_conv2d.c

```diff
@@ -0,0 +1,283 @@
+// Copyright (c) 2019, University of Washington All rights reserved.
+// 
+// Redistribution and use in source and binary forms, with or without modification,
+// are permitted provided that the following conditions are met:
+// 
+// Redistributions of source code must retain the above copyright notice, this list
+// of conditions and the following disclaimer.
+// 
+// Redistributions in binary form must reproduce the above copyright notice, this
+// list of conditions and the following disclaimer in the documentation and/or
+// other materials provided with the distribution.
+// 
+// Neither the name of the copyright holder nor the names of its contributors may
+// be used to endorse or promote products derived from this software without
+// specific prior written permission.
+// 
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+// ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+// WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+// DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR
+// ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+// (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+// LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+// ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+#include "test_conv2d.h"
+#include <math.h>
+
+#define ALLOC_NAME "default_allocator"
+
+#define MIN_N 4
+#define MAX_N 1024
+
+#define MIN_F 1
+#define MAX_F 128
+
+#define MIN_P 0
+#define MAX_P 8
+
+#define MIN_S 1
+#define MAX_S 8
+
+#define MAX_VALUE_MAGNITUDE 1e6
+
+int generate_inclusive(int min, int max)
+{
+        int range = (max - min + 1);
+        return (rand() % range) + min;
+}
+
+int output_dim(int N, int F, int P, int S)
+{
+        return 1 + (N - F + 2 * P) / S;
+}
+
+// Takes an MxN matrix A and a HxW filter, a padding P,
+// a vertical stride Sy and a horizontal Sx, and outputs
+// the result of a 2D convolution into B. B must be of size
+// output_dim(M, H, P, Sy) x (N, W, P, Sx)
+void conv2d(const float *A,
```

Comment:
Likewise for here - you've written this for float, but we're also interested in other datatypes. Writing this in C++ with templates would make it a bit more general